File: C++/Trabalho_POO_parte_2/Carro_Passeio.cpp
#include "Veiculo_Motorizado.cpp"
#include "Interface_IPVA.h"

class Carro_Passeio : public Veiculo_Motorizado, public Interface_IPVA
{
private:
    static constexpr float gasto_carro_passeio = 0.75;
    char tipo = 'P';

public:
    Carro_Passeio(int id, int rodas) : Veiculo_Motorizado(id, rodas), tipo('C')
    {
        std::cout << "Carro de Passeio com ID " << get_id() << " foi construído" << std::endl;
    }

    char get_tipo()
    {
        return tipo;
    }

    void mover()
    {
        if (get_ipva())
        {
            if (get_combustivel() >= gasto_carro_passeio)
            {
                if (verificar_pneu())
                {
                    consumir(gasto_carro_passeio);
                    int new_distancia = get_distancia_percorrida() + 2;
                    set_distancia_percorrida(new_distancia);
                    std::cout << "Carro de passeio - ID:" << get_id() << " andou  bloco +5 e está com " << get_combustivel() << " litros de gasolina" << std::endl;
                }
                else
                    std::cout << "Carro de passeio com ID " << get_id() << " não possui todos os pneus calibrados" << std::endl;
            }
            else
                std::cout << "Carro de passeio com ID " << get_id() << " não possui combustível suficiente" << std::endl;
        }
        else
            std::cout << "Carro de passeio com ID " << get_id() << " não está com IPVA em dia" << std::endl;
    }
    void desenhar()
    {
        std::string espacos(get_distancia_percorrida(), ' ');
        std::cout << espacos << "    ____" << std::endl;
        std::cout << espacos << " __/  |_ \\_" << std::endl;
        std::cout << espacos << "|  _     _``-." << std::endl;
        std::cout << espacos << "'-(_)---(_)--'" << std::endl;
    }
};

File: Python/Trabalho_POO_parte_2/Carro_Passeio.py
from Veiculo_Motorizado import *
from Veiculo import *
from Interface_IPVA import *

class Carro_Passeio(Veiculo_Motorizado, Interface_IPVA):
    __gasto_carro_passeio: float = 0.75
    def __init__(self, id: int, rodas: int):
        super().__init__(id, rodas)
        super().set_valor_ipva(self.calcular_ipva())
        self.__tipo = "C"
        print(f"Carro de Passeio com ID {self.get_id()} foi construida")

    # metodo get
    def get_tipo(self) -> int:
        return self.__tipo

    # funções abstrata herdada pelo pai veiculo
    def mover(self):
        if self.get_ipva():
            if self.get_combustivel() >= self.__gasto_carro_passeio:
                for roda in self.get_rodas():
                    if roda.verificar_pneu():
                        self.consumir(self.__gasto_carro_passeio)
                        new_distancia = self.get_distancia_percorrida() + 2
                        self.set_distancia_percorrida(new_distancia)
                        print(
                            f"Carro de passeio - ID:{self.get_id()} andou  bloco +5 e está com {self.get_combustivel():.2f}litros de gasolina")
                        return
                print(
                    f"Carro de passeio com ID {self.get_id()} não possui todos os pneus calibrados")
            else:
                print(
                    f"Carro de passeio com ID {self.get_id()} não possui combustivel suficiente")
        else:
            print(
                f"Carro de passeio com ID {self.get_id()} não está com ipva em dia")

    def desenhar(self) -> None:
        print(" "*self.get_distancia_percorrida() + "    ____")
        print(" "*self.get_distancia_percorrida() + " __/  |_ \_")
        print(" "*self.get_distancia_percorrida() + "|  _     _``-.")
        print(" "*self.get_distancia_percorrida() + "'-(_)---(_)--'")
  
    def calcular_ipva(self):
        return self.VALOR_BASE * self.CARRO_PASSEIO